src/borsh.rs

```diff
@@ -0,0 +1,124 @@
+#![cfg_attr(docsrs, doc(cfg(feature = "borsh")))]
+
+use alloc::vec::Vec;
+use core::hash::BuildHasher;
+use core::hash::Hash;
+use core::iter::ExactSizeIterator;
+use core::mem::size_of;
+
+use borsh::error::ERROR_ZST_FORBIDDEN;
+use borsh::io::{Error, ErrorKind, Read, Result, Write};
+use borsh::{BorshDeserialize, BorshSerialize};
+
+use crate::map::IndexMap;
+use crate::set::IndexSet;
+
+impl<K, V, H> BorshSerialize for IndexMap<K, V, H>
+where
+    K: BorshSerialize,
+    V: BorshSerialize,
+    H: BuildHasher,
+{
+    #[inline]
+    fn serialize<W: Write>(&self, writer: &mut W) -> Result<()> {
+        check_zst::<K>()?;
+
+        let iterator = self.iter();
+
+        u32::try_from(iterator.len())
+            .map_err(|_| ErrorKind::InvalidData)?
+            .serialize(writer)?;
+
+        for (key, value) in iterator {
+            key.serialize(writer)?;
+            value.serialize(writer)?;
+        }
+
+        Ok(())
+    }
+}
+
+impl<K, V, H> BorshDeserialize for IndexMap<K, V, H>
+where
+    K: BorshDeserialize + Eq + Hash,
+    V: BorshDeserialize,
+    H: BuildHasher + Default,
+{
+    #[inline]
+    fn deserialize_reader<R: Read>(reader: &mut R) -> Result<Self> {
+        check_zst::<K>()?;
+        let vec = <Vec<(K, V)>>::deserialize_reader(reader)?;
```

Comment:
I see that borsh cites performance when it uses an intermediate vec like this for other maps, but I wonder if that's worthwhile in this case where we're ultimately storing map entries in a vec anyway. Could you try with our own deserializing loop and compare it?

Reply:
I wrote a crude bench to test deserialization:
```rs
#[bench]
fn borsh_deserialize(b: &mut Bencher) {
    use borsh::BorshDeserialize;

    let serialized_map = borsh::to_vec(&*IMAP_100K).unwrap();

    b.iter(|| {
        let _: IndexMap<u32, u32> = BorshDeserialize::try_from_slice(&serialized_map).unwrap();
    });
}
```

with `Vec::deserialize_reader` (the current code), here are a couple of `cargo bench` runs:
```
test borsh_deserialize                     ... bench:   2,536,210 ns/iter (+/- 43,823)
test borsh_deserialize                     ... bench:   2,536,983 ns/iter (+/- 45,303)
test borsh_deserialize                     ... bench:   2,545,633 ns/iter (+/- 69,611)
```

with the following changes:
```diff
diff --git a/src/borsh.rs b/src/borsh.rs
index e0bab1a..85a8d9b 100644
--- a/src/borsh.rs
+++ b/src/borsh.rs
@@ -47,8 +47,17 @@ where
     #[inline]
     fn deserialize_reader<R: Read>(reader: &mut R) -> Result<Self> {
         check_zst::<K>()?;
-        let vec = <Vec<(K, V)>>::deserialize_reader(reader)?;
-        Ok(vec.into_iter().collect::<IndexMap<K, V, H>>())
+
+        let elems = u32::deserialize_reader(reader)?;
+        let mut map = Self::with_capacity_and_hasher(elems as _, Default::default());
+
+        for _ in 0..elems {
+            let key = K::deserialize_reader(reader)?;
+            let value = V::deserialize_reader(reader)?;
+            map.insert(key, value);
+        }
+
+        Ok(map)
     }
 }
```
we get the following results:
```
test borsh_deserialize                     ... bench:   2,534,810 ns/iter (+/- 172,217)
test borsh_deserialize                     ... bench:   2,362,186 ns/iter (+/- 153,488)
test borsh_deserialize                     ... bench:   2,533,979 ns/iter (+/- 173,925)
test borsh_deserialize                     ... bench:   2,378,631 ns/iter (+/- 163,738)
test borsh_deserialize                     ... bench:   2,532,943 ns/iter (+/- 171,355)
```

for some reason this code produces way more volatile results, but in the worst case, they're still marginally better than `Vec::deserialize_reader`, and in the best case, they're a bit better.

I'll push the new deserialization code to this branch